File: apps/mobile_ios/test/route_geometry_test.dart
import 'package:core_models/core_models.dart' show Waypoint;
import 'package:flutter_test/flutter_test.dart';

import '../lib/route_geometry.dart';

/// Pure-helper coverage for `interpolateAlongRoute` — the math that
/// drives the route-detail screen's scrubber. Drag the slider 0 → 1
/// and this helper produces the lat/lng for the runner pulse so the
/// user can preview the direction of the route step-by-step.
///
/// All test routes anchor at lat=0 so equirectangular projection
/// math (cos(lat) = 1) makes the haversine + linear interpolation
/// predictable to a sub-metre precision.
void main() {
  // 1° longitude at the equator ≈ 111 320 m. Useful for converting
  // metres ↔ longitude in tests where lat = 0.
  const metresPerDegLngAtEquator = 111320.0;

  Waypoint wp(double lat, double lng) => Waypoint(lat: lat, lng: lng);

  group('interpolateAlongRoute — guard rails', () {
    test('null on empty waypoints', () {
      expect(interpolateAlongRoute(const [], 0.5), isNull);
    });

    test('null on single waypoint', () {
      expect(
        interpolateAlongRoute([wp(0, 0)], 0.5),
        isNull,
        reason: 'A single point isn\'t a polyline — nothing to '
            'interpolate. Caller (route_detail screen) hides the '
            'scrubber entirely when waypoints.length < 2.',
      );
    });

    test(
        'all-coincident waypoints (degenerate, zero total length) '
        'snap to the first waypoint',
        () {
      // Defence-in-depth: the route builder\'s 5 m dedupe should
      // make this impossible in practice, but if it ever surfaces
      // we return a valid answer rather than dividing by zero.
      final out = interpolateAlongRoute(
        [wp(1, 1), wp(1, 1), wp(1, 1)],
        0.5,
      );
      expect(out, isNotNull);
      expect(out!.lat, 1);
      expect(out.lng, 1);
    });

    test('fraction below 0 clamps to start', () {
      final out = interpolateAlongRoute(
        [wp(0, 0), wp(0, 0.001)],
        -1.0,
      );
      expect(out!.lng, closeTo(0.0, 1e-9));
    });

    test('fraction above 1 clamps to end', () {
      final out = interpolateAlongRoute(
        [wp(0, 0), wp(0, 0.001)],
        2.0,
      );
      expect(out!.lng, closeTo(0.001, 1e-9));
    });
  });

  group('interpolateAlongRoute — fraction → position', () {
    test('fraction = 0.0 returns the start waypoint exactly', () {
      final out = interpolateAlongRoute(
        [wp(0, 0), wp(0, 0.005), wp(0, 0.010)],
        0.0,
      );
      expect(out!.lat, 0);
      expect(out.lng, closeTo(0, 1e-9));
    });

    test('fraction = 1.0 returns the end waypoint exactly', () {
      final out = interpolateAlongRoute(
        [wp(0, 0), wp(0, 0.005), wp(0, 0.010)],
        1.0,
      );
      expect(out!.lat, 0);
      expect(out.lng, closeTo(0.010, 1e-9));
    });

    test('fraction = 0.5 on an even-spaced polyline lands at midpoint',
        () {
      // Two segments of equal length → fraction 0.5 lands exactly at
      // the middle waypoint.
      final out = interpolateAlongRoute(
        [wp(0, 0), wp(0, 0.005), wp(0, 0.010)],
        0.5,
      );
      expect(out!.lng, closeTo(0.005, 1e-6));
    });

    test('distance-weighted — long-segment dominates the scrub range',
        () {
      // Two segments: short (1 unit) + long (9 units). Fraction 0.5
      // of total distance (5 units) lands INSIDE the long segment
      // — NOT at the corner waypoint — proving distance-weighted
      // interpolation. A naive index-based interpolation would
      // mistakenly snap to the corner here.
      final out = interpolateAlongRoute(
        [wp(0, 0), wp(0, 0.001), wp(0, 0.010)],
        0.5,
      );
      // 50% of total 10-unit distance = position 5 → 4 units into
      // the long second segment → lng ≈ 0.001 + (4/9 * 0.009) ≈ 0.005.
      expect(out!.lng, closeTo(0.005, 1e-6));
    });

    test('fraction = 0.25 on a 4-equal-leg polyline lands at 1st corner',
        () {
      // Four equal segments, fraction 0.25 lands exactly at the
      // first interior waypoint (3 corners + start + end = 4 legs).
      final out = interpolateAlongRoute(
        [wp(0, 0), wp(0, 0.001), wp(0, 0.002), wp(0, 0.003), wp(0, 0.004)],
        0.25,
      );
      expect(out!.lng, closeTo(0.001, 1e-6));
    });

    test('non-monotonic / out-and-back polyline interpolates by path '
        'distance, not chord', () {
      // Out + back along the same axis. Path distance: 2 units.
      // Fraction 0.5 lands at the "turn-around" point, NOT the
      // chord midpoint (which would be back at the start).
      final out = interpolateAlongRoute(
        [wp(0, 0), wp(0, 0.001), wp(0, 0)],
        0.5,
      );
      expect(
        out!.lng,
        closeTo(0.001, 1e-6),
        reason: 'Path-distance interpolation: the runner has '
            'reached the turn-around halfway through an out-and-back, '
            'NOT averaged to the chord midpoint (which would erase '
            'the loop entirely).',
      );
    });
  });

  group('interpolateAlongRoute — elevation lerp', () {
    test('elevation interpolates linearly between adjacent waypoints',
        () {
      final out = interpolateAlongRoute(
        [
          Waypoint(lat: 0, lng: 0, elevationMetres: 100),
          Waypoint(lat: 0, lng: 0.001, elevationMetres: 200),
        ],
        0.5,
      );
      expect(
        out!.elevationMetres,
        closeTo(150, 0.01),
        reason: 'Halfway between 100 m and 200 m elevation should '
            'lerp to 150 m so the marker\'s tooltip / readout can '
            'show smooth elevation as the user drags.',
      );
    });

    test('elevation lerp tolerates one-sided null (a/null or null/b)',
        () {
      // Only the END waypoint carries elevation — interpolation
      // returns the populated value rather than crashing.
      final out = interpolateAlongRoute(
        [
          Waypoint(lat: 0, lng: 0),
          Waypoint(lat: 0, lng: 0.001, elevationMetres: 50),
        ],
        0.5,
      );
      expect(out!.elevationMetres, 50);
    });

    test('elevation lerp returns null when both sides are null', () {
      final out = interpolateAlongRoute(
        [wp(0, 0), wp(0, 0.001)],
        0.5,
      );
      expect(out!.elevationMetres, isNull);
    });
  });

  group('polylineLengthMetres', () {
    test('empty / single-point polyline → 0', () {
      expect(polylineLengthMetres(const []), 0);
      expect(polylineLengthMetres([wp(0, 0)]), 0);
    });

    test('single 100-m segment → roughly 100 m at the equator', () {
      // 100 m / 111 320 m/° ≈ 8.98e-4° of longitude.
      final out = polylineLengthMetres([
        wp(0, 0),
        wp(0, 100 / metresPerDegLngAtEquator),
      ]);
      expect(out, closeTo(100, 1));
    });

    test('multi-segment lengths sum', () {
      // Two 100-m segments → 200 m total.
      final step = 100 / metresPerDegLngAtEquator;
      final out = polylineLengthMetres([
        wp(0, 0),
        wp(0, step),
        wp(0, 2 * step),
      ]);
      expect(out, closeTo(200, 1));
    });
  });

  group('interpolateAlongRoute — geographic + edge cases', () {
    test('southern-hemisphere polyline interpolates symmetrically '
        'to a northern-hemisphere mirror', () {
      // Pin that the helper isn\'t accidentally relying on a
      // positive-lat assumption (e.g. a sign error in the haversine).
      // South-of-equator interp at fraction=0.5 lands at -0.005°
      // exactly, mirroring the north case.
      final southOut = interpolateAlongRoute(
        [wp(0, 0), wp(-0.005, 0), wp(-0.010, 0)],
        0.5,
      );
      expect(southOut!.lat, closeTo(-0.005, 1e-6));
      expect(southOut.lng, 0);
    });

    test(
        'eastern + western longitude interp is symmetric (negative-lng safe)',
        () {
      // Same defence for negative longitudes — e.g. polylines in
      // the Americas. fraction=0.5 of [0,-0.010] → -0.005.
      final out = interpolateAlongRoute(
        [wp(0, 0), wp(0, -0.005), wp(0, -0.010)],
        0.5,
      );
      expect(out!.lng, closeTo(-0.005, 1e-6));
    });

    test('2-waypoint polyline at fraction = 0.5 lands exactly at midpoint',
        () {
      // Minimal valid input — the scrubber should still work on a
      // 2-pin route (the route builder\'s minimum-renderable state).
      final out = interpolateAlongRoute(
        [wp(0, 0), wp(0, 0.010)],
        0.5,
      );
      expect(out!.lng, closeTo(0.005, 1e-6));
    });

    test(
        'segments of zero length are skipped — adjacent duplicate '
        'waypoints don\'t poison interpolation',
        () {
      // After the 5-m dedupe guard fires in route_builder, a saved
      // route should never carry exact duplicates. But defensively
      // — if it does, the helper must skip the zero-length leg
      // (segLen <= 0 continue) and continue into the next real
      // segment. Otherwise fraction=0.5 on [(0,0), (0,0), (0,0.01)]
      // would return (0,0) (incorrect).
      final out = interpolateAlongRoute(
        [wp(0, 0), wp(0, 0), wp(0, 0.010)],
        0.5,
      );
      expect(
        out!.lng,
        closeTo(0.005, 1e-6),
        reason: 'Helper must skip the zero-length first leg and '
            'land at the midpoint of the real 0→0.010 leg.',
      );
    });

    test(
        'long polyline (200 waypoints) is O(n) — runs under 50 ms',
        () {
      // The recorder + route builder produce polylines bounded by
      // ~1000 points; pin the helper at 200 to catch a quadratic
      // regression early.
      final wps = [
        for (var i = 0; i <= 200; i++) wp(0, i * 0.0001),
      ];
      final sw = Stopwatch()..start();
      final out = interpolateAlongRoute(wps, 0.5);
      sw.stop();
      expect(out, isNotNull);
      expect(
        sw.elapsedMilliseconds,
        lessThan(50),
        reason: 'Linear scan over 200 points must finish well under '
            '50 ms — pin against a quadratic refactor.',
      );
    });
  });
}

File: apps/web/src/lib/routes/route_geometry.test.ts
// Unit tests for the route-geometry helpers. Run with `node --test`
// via tsx, matching the rest of the apps/web suite:
//   npx tsx --test src/lib/routes/route_geometry.test.ts
//
// Mirror of `apps/mobile_android/test/route_geometry_test.dart` —
// keep this file in lockstep with the Dart twin.

import { test } from 'node:test';
import assert from 'node:assert/strict';

import {
	interpolateAlongRoute,
	distanceAlongRoute,
	polylineLengthMetres,
	type RouteWaypoint,
} from './route_geometry';

const wp = (lat: number, lng: number, elev?: number): RouteWaypoint => ({
	lat,
	lng,
	elevation_m: elev,
});

test('interpolateAlongRoute — null on empty waypoints', () => {
	assert.equal(interpolateAlongRoute([], 0.5), null);
});

test('interpolateAlongRoute — null on single waypoint', () => {
	assert.equal(interpolateAlongRoute([wp(0, 0)], 0.5), null);
});

test('interpolateAlongRoute — all-coincident waypoints snap to start', () => {
	const out = interpolateAlongRoute(
		[wp(1, 1), wp(1, 1), wp(1, 1)],
		0.5,
	);
	assert.ok(out);
	assert.equal(out!.lat, 1);
	assert.equal(out!.lng, 1);
});

test('interpolateAlongRoute — fraction below 0 clamps to start', () => {
	const out = interpolateAlongRoute([wp(0, 0), wp(0, 0.001)], -1.0);
	assert.ok(out);
	assert.ok(Math.abs(out!.lng - 0) < 1e-9);
});

test('interpolateAlongRoute — fraction above 1 clamps to end', () => {
	const out = interpolateAlongRoute([wp(0, 0), wp(0, 0.001)], 2.0);
	assert.ok(out);
	assert.ok(Math.abs(out!.lng - 0.001) < 1e-9);
});

test('interpolateAlongRoute — fraction = 0 returns start exactly', () => {
	const out = interpolateAlongRoute(
		[wp(0, 0), wp(0, 0.005), wp(0, 0.010)],
		0.0,
	);
	assert.equal(out!.lat, 0);
	assert.ok(Math.abs(out!.lng - 0) < 1e-9);
});

test('interpolateAlongRoute — fraction = 1 returns end exactly', () => {
	const out = interpolateAlongRoute(
		[wp(0, 0), wp(0, 0.005), wp(0, 0.010)],
		1.0,
	);
	assert.equal(out!.lat, 0);
	assert.ok(Math.abs(out!.lng - 0.010) < 1e-9);
});

test('interpolateAlongRoute — fraction = 0.5 on even-spaced lands at midpoint', () => {
	const out = interpolateAlongRoute(
		[wp(0, 0), wp(0, 0.005), wp(0, 0.010)],
		0.5,
	);
	assert.ok(Math.abs(out!.lng - 0.005) < 1e-6);
});

test('interpolateAlongRoute — distance-weighted (long segment dominates)', () => {
	// Short (1 unit) + long (9 units). Fraction 0.5 of total
	// distance (5 units) lands INSIDE the long segment — not at
	// the corner. Catches a regression to naive index-weighted.
	const out = interpolateAlongRoute(
		[wp(0, 0), wp(0, 0.001), wp(0, 0.010)],
		0.5,
	);
	assert.ok(Math.abs(out!.lng - 0.005) < 1e-6);
});

test('interpolateAlongRoute — 4-equal-leg at 0.25 lands at first corner', () => {
	const out = interpolateAlongRoute(
		[wp(0, 0), wp(0, 0.001), wp(0, 0.002), wp(0, 0.003), wp(0, 0.004)],
		0.25,
	);
	assert.ok(Math.abs(out!.lng - 0.001) < 1e-6);
});

test('interpolateAlongRoute — out-and-back uses path distance, not chord', () => {
	const out = interpolateAlongRoute(
		[wp(0, 0), wp(0, 0.001), wp(0, 0)],
		0.5,
	);
	// At the turn-around, NOT at the chord midpoint.
	assert.ok(Math.abs(out!.lng - 0.001) < 1e-6);
});

test('interpolateAlongRoute — elevation lerps linearly', () => {
	const out = interpolateAlongRoute(
		[wp(0, 0, 100), wp(0, 0.001, 200)],
		0.5,
	);
	assert.ok(Math.abs((out!.elevation_m ?? 0) - 150) < 0.01);
});

test('interpolateAlongRoute — lerp tolerates one-sided null', () => {
	const out = interpolateAlongRoute(
		[wp(0, 0), wp(0, 0.001, 50)],
		0.5,
	);
	assert.equal(out!.elevation_m, 50);
});

test('interpolateAlongRoute — returns null elevation when both sides null', () => {
	const out = interpolateAlongRoute([wp(0, 0), wp(0, 0.001)], 0.5);
	assert.equal(out!.elevation_m, null);
});

test('polylineLengthMetres — empty / single → 0', () => {
	assert.equal(polylineLengthMetres([]), 0);
	assert.equal(polylineLengthMetres([wp(0, 0)]), 0);
});

test('polylineLengthMetres — 100 m segment at equator ≈ 100 m', () => {
	const metresPerDegLngAtEquator = 111320.0;
	const out = polylineLengthMetres([
		wp(0, 0),
		wp(0, 100 / metresPerDegLngAtEquator),
	]);
	assert.ok(Math.abs(out - 100) < 1);
});

test('polylineLengthMetres — multi-segment lengths sum', () => {
	const metresPerDegLngAtEquator = 111320.0;
	const step = 100 / metresPerDegLngAtEquator;
	const out = polylineLengthMetres([wp(0, 0), wp(0, step), wp(0, 2 * step)]);
	assert.ok(Math.abs(out - 200) < 1);
});

test('interpolateAlongRoute — southern-hemisphere is symmetric', () => {
	const out = interpolateAlongRoute(
		[wp(0, 0), wp(-0.005, 0), wp(-0.010, 0)],
		0.5,
	);
	assert.ok(Math.abs(out!.lat - -0.005) < 1e-6);
	assert.equal(out!.lng, 0);
});

test('interpolateAlongRoute — negative longitude (Americas) is safe', () => {
	const out = interpolateAlongRoute(
		[wp(0, 0), wp(0, -0.005), wp(0, -0.010)],
		0.5,
	);
	assert.ok(Math.abs(out!.lng - -0.005) < 1e-6);
});

test('interpolateAlongRoute — 2-waypoint polyline at fraction=0.5 → midpoint', () => {
	// Minimum valid input — pin the smallest case the scrubber must
	// support.
	const out = interpolateAlongRoute([wp(0, 0), wp(0, 0.010)], 0.5);
	assert.ok(Math.abs(out!.lng - 0.005) < 1e-6);
});

test('interpolateAlongRoute — skips zero-length segments (no poison from duplicate waypoints)', () => {
	// Defence-in-depth: the route builder\'s 5-m dedupe should
	// guarantee no exact duplicates, but if they leak through the
	// helper must skip and land in the next real segment.
	const out = interpolateAlongRoute(
		[wp(0, 0), wp(0, 0), wp(0, 0.010)],
		0.5,
	);
	assert.ok(Math.abs(out!.lng - 0.005) < 1e-6);
});

test('interpolateAlongRoute — 200-point polyline runs under 50 ms (O(n) guard)', () => {
	const wps: RouteWaypoint[] = [];
	for (let i = 0; i <= 200; i++) wps.push(wp(0, i * 0.0001));
	const start = performance.now();
	const out = interpolateAlongRoute(wps, 0.5);
	const elapsed = performance.now() - start;
	assert.ok(out !== null);
	assert.ok(
		elapsed < 50,
		`Expected <50ms, got ${elapsed.toFixed(1)}ms — quadratic regression?`,
	);
});

const metresPerDegLng = 111320.0;
const distWp = (lng: number): RouteWaypoint => wp(0, lng / metresPerDegLng);

test('distanceAlongRoute — null on < 2 waypoints', () => {
	assert.equal(distanceAlongRoute({ lat: 0, lng: 0 }, []), null);
	assert.equal(distanceAlongRoute({ lat: 0, lng: 0 }, [wp(0, 0)]), null);
});

test('distanceAlongRoute — point on a vertex returns its cumulative distance', () => {
	// Three 100-m legs along the equator. The 2nd vertex is at 200 m.
	const wps = [distWp(0), distWp(100), distWp(200), distWp(300)];
	const d = distanceAlongRoute(wps[2], wps);
	assert.ok(d !== null);
	assert.ok(Math.abs(d! - 200) < 1, `got ${d}`);
});

test('distanceAlongRoute — point mid-segment returns the interpolated distance', () => {
	const wps = [distWp(0), distWp(100), distWp(200)];
	const d = distanceAlongRoute(distWp(150), wps);
	assert.ok(d !== null);
	assert.ok(Math.abs(d! - 150) < 1, `got ${d}`);
});

test('distanceAlongRoute — perpendicular offset still maps to the right along-distance', () => {
	// 50 m north of the 150-m mark — projects back down to 150 m.
	const wps = [distWp(0), distWp(100), distWp(200)];
	const offset = { lat: 50 / metresPerDegLng, lng: 150 / metresPerDegLng };
	const d = distanceAlongRoute(offset, wps);
	assert.ok(d !== null);
	assert.ok(Math.abs(d! - 150) < 1, `got ${d}`);
});

test('distanceAlongRoute — point near the end maps near totalLength', () => {
	const wps = [distWp(0), distWp(100), distWp(200)];
	const total = polylineLengthMetres(wps);
	const d = distanceAlongRoute(distWp(199), wps);
	assert.ok(d !== null);
	assert.ok(Math.abs(d! - 199) < 1, `got ${d}`);
	assert.ok(d! <= total + 1e-6);
});

test('distanceAlongRoute — picks the nearest of two close segments', () => {
	// An L: east 100 m then north 100 m. A point just east of the
	// corner, slightly north, is nearest the FIRST (horizontal) leg,
	// so it maps to ~100 m, not into the vertical leg.
	const corner = distWp(100);
	const up = wp(100 / metresPerDegLng, 100 / metresPerDegLng);
	const wps = [distWp(0), corner, up];
	// Just south of the 50-m mark on the first (horizontal) leg —
	// unambiguously nearest it, far from the vertical leg.
	const probe = { lat: -2 / metresPerDegLng, lng: 50 / metresPerDegLng };
	const d = distanceAlongRoute(probe, wps);
	assert.ok(d !== null);
	assert.ok(d! < 100, `expected on the first leg (<100 m), got ${d}`);
	assert.ok(Math.abs(d! - 50) < 2, `got ${d}`);
});

test('distanceAlongRoute — clamps to [0, totalLength]', () => {
	const wps = [distWp(0), distWp(100), distWp(200)];
	const total = polylineLengthMetres(wps);
	// Way past the end, off to the side.
	const far = distWp(10_000);
	const d = distanceAlongRoute(far, wps);
	assert.ok(d !== null);
	assert.ok(d! >= 0 && d! <= total + 1e-6, `got ${d} (total ${total})`);
});
